File: ifoodAndroidTest/app/src/main/java/test/ifood/uellisson/ifoodandroidtest/view/TweetAdapter.java
package test.ifood.uellisson.ifoodandroidtest.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import test.ifood.uellisson.ifoodandroidtest.R;
import test.ifood.uellisson.ifoodandroidtest.model.Tweet;

public class TweetAdapter extends RecyclerView.Adapter<TweetViewHolder>{
    Context context;
    List<Tweet> tweetList;

    public TweetAdapter(Context context, List<Tweet> tweetList) {
        super();
        this.context = context;
        this.tweetList = tweetList;
    }


    @NonNull
    @Override
    public TweetViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(context).inflate(R.layout.tweet_item, viewGroup, false);
        return new TweetViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull TweetViewHolder tweetViewHolder, int i) {
        tweetViewHolder.tweetMessage.setText(tweetList.get(i).getTwitterMessage());
        tweetViewHolder.crateAt.setText(tweetList.get(i).getCreateAt());
    }

    @Override
    public int getItemCount() {
        return tweetList.size();
    }
}

File: twittermoodanalyzer/app/src/main/java/com/drury/twittermoodanalyzer/view/adapter/TweetAdapter.kt
package com.drury.twittermoodanalyzer.view.adapter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.drury.twittermoodanalyzer.R
import com.drury.twittermoodanalyzer.extension.toSimpleString
import com.drury.twittermoodanalyzer.model.TweetModel
import com.drury.twittermoodanalyzer.utils.AppConstants
import com.drury.twittermoodanalyzer.view.MoodActivity
import kotlinx.android.synthetic.main.tweet_item.view.*
import java.util.*

class TweetAdapter(var dataset: MutableList<TweetModel>, var activity: Activity) :
        RecyclerView.Adapter<TweetAdapter.TweetViewHolder>() {

    class TweetViewHolder(val view: View) : RecyclerView.ViewHolder(view) {

        fun bind(tweet: TweetModel) {
            view.textViewTweet.text = tweet.text
            view.textViewDate.text = Date(tweet.created).toSimpleString()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): TweetAdapter.TweetViewHolder {

        val view = LayoutInflater.from(activity)
                .inflate(R.layout.tweet_item, parent, false)

        return TweetViewHolder(view).listen { pos, type ->
            val item = dataset[pos]
            val intent = Intent(activity, MoodActivity::class.java).putExtra(AppConstants.TWEET, item)
            activity.startActivity(intent)
        }
    }

    override fun onBindViewHolder(holder: TweetViewHolder, position: Int) {
        holder.bind(dataset[position])
    }

    override fun getItemCount() = dataset.size
}

fun <T : RecyclerView.ViewHolder> T.listen(event: (position: Int, type: Int) -> Unit): T {
    itemView.setOnClickListener {
        event.invoke(adapterPosition, itemViewType)
    }
    return this
}